text/0000-read-all.md

```diff
@@ -0,0 +1,50 @@
+- Feature Name: read_all
+- Start Date: 2015-03-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Rust's Write trait has write_all, which attempts to write an entire
+buffer.  This proposal adds read_all, which attempts to read a fixed
+number of bytes into a given buffer.
+
+# Motivation
+
+The new read_all method will allow programs to read from disk without
+having to write their own read loops.  Most Rust programs which need
+to read from disk will prefer this to the plain read function.  Many C
+programs have the same need, and solve it the same way (e.g. git has
+read_in_full).  Here's one example of a Rust library doing this:
+https://github.com/BurntSushi/byteorder/blob/master/src/new.rs#L184
+
+# Detailed design
+
+The read_all function will take a mutable, borrowed slice of u8 to
+read into, and will attempt to fill that entire slice with data.
+
+It will loop, calling read() once per iteration and attempting to read
+the remaining amount of data.  If read returns EINTR, the loop will
+retry.  If there are no more bytes to read (as signalled by a return
+of Ok(0) from read()), a new error type, ErrorKind::ReadZero, will be
+returned.  In the event of another error, that error will be
+returned. After a read call returns having successfully read some
+bytes, the total number of bytes read will be updated.  If that
+total is equal to the size of the buffer, read will return
+successfully.
```

Comment:
@alexcrichton, the second case (Some bytes were read, and then we saw `Ok(0)`) is the only case for which I would generally care about knowing how many bytes were read. For the I/O error case, the vast majority of the time I don't care about how many bytes were read, I just bail out and return the error. I'm much more likely to care in the `Ok(0)` case, such as if I'm reading fixed-size records from a file or the network, but the last record is allowed to be short.

Since the EOF condition doesn't exist for `write_all`, no changes would be needed there for consistency.

One way to implement this would be as follows:
* Add `ShortRead(usize)` to `Repr`
* Also add `ShortRead(usize)` to `ErrorKind`
* Add `Repr::ShortRead(n) => ErrorKind::ShortRead(n)` to `Error::kind` (and also update `description` and `detail`)

This would not increase the size of `io::Error`. (`io::Error` already has `Custom(Box<Custom>)`, which is the same size). It would increase the size of `Custom`, but that always gets boxed. It would also allow the return value to stay `io::Result<()>`, as desired, and would maintain usability by keeping the check for a short read out of the normal program flow and allowing a short read to be treated as a normal error value (via `try!`, et cetera) unless the user explicitly wants to handle it differently.